File: homework/Truong_Duong/Lesson_7/112_Path_Sum.js
// Solution for https://leetcode.com/problems/path-sum/

var hasPathSum = function(root, targetSum) {
    if(root === null) return false;
    if(root.left === null && root.right === null) return root.val === targetSum;
    
    const left = hasPathSum(root.left, targetSum - root.val);
    const right = hasPathSum(root.right, targetSum - root.val);
    return left || right;
};

File: homework/thinhbka/lesson-5-graph/112._Path_Sum.py
# https://leetcode.com/problems/path-sum/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self,node,target):
        if node is None:return False
        if node.left is None and node.right is None: # node is a leaf
            return node.val == target
        return self.dfs(node.left,target-node.val) or self.dfs(node.right,target-node.val) 
    
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return self.dfs(root,targetSum)